File: server/routes/bottles.js
const express = require('express');
const router = express.Router();
const database = require('../database');
const multer = require('multer');
const path = require('path');
const fs = require('fs');
const mistralAnalyzer = require('../mistralAnalyzer');
const mistralAI = require('../mistralAI');
const mistralConfig = require('../mistralConfig');

const ai = mistralAI;

const upload = multer({
    storage: multer.diskStorage({
        destination: (req, file, cb) => {
            const uploadDir = path.join(__dirname, '../../public/uploads');
            if (!fs.existsSync(uploadDir)) fs.mkdirSync(uploadDir, { recursive: true });
            cb(null, uploadDir);
        },
        filename: (req, file, cb) => {
            const uniqueSuffix = Date.now() + '-' + Math.round(Math.random() * 1E9);
            cb(null, uniqueSuffix + path.extname(file.originalname));
        },
        limits: { fileSize: 5 * 1024 * 1024 }
    })
});

router.get('/', (req, res) => {
    database.getAllBottles((err, bottles) => {
        if (err) res.status(500).json({ error: "Erreur lors de la récupération des bouteilles" });
        else res.json(bottles);
    });
});

router.post('/analyze', upload.single('image'), async (req, res) => {
    try {
        let imagePath = null;
        if (req.file) imagePath = path.join(__dirname, '../../public/uploads', req.file.filename);
        const result = await mistralAnalyzer.analyzeBottleWithMistralOnly(imagePath, false);
        res.json({
            success: true,
            bottleInfo: result,
            mistralAvailable: !!mistralConfig.apiKey,
            googleVisionAvailable: !!process.env.GOOGLE_VISION_API_KEY,
            analysisMethod: result.analysisMethod,
            extractedText: result.extractedText
        });
    } catch (error) {
        console.error("Erreur analyse IA :", error);
        res.status(500).json({
            success: false,
            error: "Erreur serveur lors de l'analyse",
            mistralAvailable: !!mistralConfig.apiKey,
            googleVisionAvailable: !!process.env.GOOGLE_VISION_API_KEY,
            bottleInfo: mistralAnalyzer.getFallbackBottleInfo(error.message)
        });
    }
});

// NOUVEAU : Route pour l'analyse en 2 étapes (Google Vision → Mistral)
router.post('/analyze-two-step', upload.single('image'), async (req, res) => {
    try {
        let imagePath = null;
        if (req.file) imagePath = path.join(__dirname, '../../public/uploads', req.file.filename);
        
        // Récupérer les données manuelles si fournies
        const manualData = req.body.name || req.body.year ? {
            name: req.body.name || null,
            year: req.body.year ? parseInt(req.body.year) : null
        } : null;
        
        const result = await mistralAnalyzer.analyzeBottleWithTwoStepProcess(imagePath, false, manualData);
        
        res.json({
            success: true,
            bottleInfo: result,
            mistralAvailable: !!mistralConfig.apiKey,
            googleVisionAvailable: !!process.env.GOOGLE_VISION_API_KEY,
            analysisMethod: result.analysisMethod,
            extractedText: result.extractedText,
            requiresManualInput: result.requiresManualInput || false,
            partialData: result.partialData || null
        });
    } catch (error) {
        console.error("Erreur analyse IA 2 étapes :", error);
        res.status(500).json({
            success: false,
            error: "Erreur serveur lors de l'analyse",
            mistralAvailable: !!mistralConfig.apiKey,
            googleVisionAvailable: !!process.env.GOOGLE_VISION_API_KEY,
            bottleInfo: mistralAnalyzer.getFallbackBottleInfo(error.message),
            requiresManualInput: true,
            partialData: null
        });
    }
});

router.post('/analyze-text', async (req, res) => {
    try {
        const { name, year, grapes, region } = req.body;
        if (!name && !year && !grapes && !region) {
            return res.status(400).json({ success: false, error: "Au moins un champ est requis" });
        }
        const manualData = { name, year, grapes, region };
        const result = await mistralAnalyzer.analyzeWithManualText(manualData);
        res.json({ success: true, bottleInfo: result, analysisMethod: result.analysisMethod || "Analyse texte manuel" });
    } catch (error) {
        console.error("Erreur analyse texte :", error);
        res.status(500).json({ success: false, error: "Erreur serveur", bottleInfo: mistralAnalyzer.getFallbackBottleInfo(error.message) });
    }
});

router.post('/analyze-base64', async (req, res) => {
    try {
        const { image } = req.body;
        if (!image) return res.status(400).json({ success: false, error: "Aucune image fournie" });
        const result = await mistralAnalyzer.analyzeBottleWithMistralOnly(image, true);
        res.json({ success: true, bottleInfo: result, mistralAvailable: !!mistralConfig.apiKey, analysisMethod: result.analysisMethod });
    } catch (error) {
        console.error("Erreur analyse base64 :", error);
        res.status(500).json({ success: false, error: "Erreur serveur", mistralAvailable: !!mistralConfig.apiKey, bottleInfo: mistralAnalyzer.getFallbackBottleInfo(error.message) });
    }
});

router.post('/', upload.single('photo'), (req, res) => {
    const bottleData = req.body;
    if (req.file) bottleData.photo = req.file.filename;
    if (bottleData.photo && bottleData.photo.startsWith('data:image/')) {
        // CORRECTION: regex avec backslashes correctement échappés
        const base64Data = bottleData.photo.replace(/^data:image\/\w+;base64,/, '');
        const buffer = Buffer.from(base64Data, 'base64');
        const uploadDir = path.join(__dirname, '../../public/uploads');
        if (!fs.existsSync(uploadDir)) fs.mkdirSync(uploadDir, { recursive: true });
        const filename = Date.now() + '-camera.jpg';
        const filePath = path.join(uploadDir, filename);
        fs.writeFileSync(filePath, buffer);
        bottleData.photo = filename;
    }
    database.saveBottle(bottleData, (err) => {
        if (err) res.status(500).json({ error: "Erreur lors de la sauvegarde" });
        else res.json({ success: true });
    });
});

router.delete('/', (req, res) => {
    const { row, col } = req.query;
    database.deleteBottle(row, col, (err) => {
        if (err) res.status(500).json({ error: "Erreur lors de la suppression" });
        else res.json({ success: true });
    });
});

router.get('/recommendations', (req, res) => {
    const { occasion, food, budget } = req.query;
    const recommendations = mistralAI.recommendWineForOccasion(occasion, food, budget);
    res.json({ success: true, recommendations });
});

router.get('/search', (req, res) => {
    const { name } = req.query;
    if (!name) return res.status(400).json({ success: false, error: "Le nom du vin est requis" });
    const wineInfo = mistralAI.searchWineByName(name);
    res.json({ success: wineInfo ? true : false, wine: wineInfo });
});

router.post('/reset-db', (req, res) => {
    try {
        const dbPath = path.join(__dirname, '../../data/cave.db');
        if (fs.existsSync(dbPath)) fs.unlinkSync(dbPath);
        res.json({ success: true, message: "Base de données réinitialisée" });
    } catch (error) {
        console.error("Erreur réinitialisation DB :", error);
        res.status(500).json({ success: false, error: "Erreur lors de la réinitialisation" });
    }
});

router.post('/chat', async (req, res) => {
    try {
        const { question } = req.body;
        if (!question) return res.status(400).json({ success: false, error: "Aucune question fournie" });
        const response = await mistralAI.askChatQuestion(question);
        res.json({ success: true, response: response, model: mistralConfig.model });
    } catch (error) {
        console.error("Erreur chat IA :", error);
        res.status(500).json({ success: false, error: "Erreur serveur lors de la requête IA" });
    }
});

router.post('/wine-pairing', async (req, res) => {
    try {
        const { food } = req.body;
        if (!food) return res.status(400).json({ success: false, error: "Aucun plat spécifié" });
        const advice = await mistralAI.getWinePairingAdvice(food);
        res.json({ success: true, advice: advice });
    } catch (error) {
        console.error("Erreur accords mets-vins :", error);
        res.status(500).json({ success: false, error: "Erreur serveur" });
    }
});

router.post('/wine-info', async (req, res) => {
    try {
        const { wineName } = req.body;
        if (!wineName) return res.status(400).json({ success: false, error: "Aucun nom de vin spécifié" });
        const info = await mistralAI.getWineInfo(wineName);
        res.json({ success: true, info: info });
    } catch (error) {
        console.error("Erreur infos vin :", error);
        res.status(500).json({ success: false, error: "Erreur serveur" });
    }
});

router.get('/mistral-status', (req, res) => {
    const hasMistralApiKey = !!mistralConfig.apiKey;
    const hasGoogleVisionApiKey = !!process.env.GOOGLE_VISION_API_KEY;
    const model = mistralConfig.model || 'mistral-tiny';
    res.json({
        success: true,
        mistralAvailable: hasMistralApiKey,
        googleVisionAvailable: hasGoogleVisionApiKey,
        model: model,
        message: hasMistralApiKey && hasGoogleVisionApiKey ?
            "Mistral AI et Google Vision sont configurés - Analyse complète disponible" :
            hasMistralApiKey ?
                "Mistral AI configuré, mais Google Vision non configuré (OCR limité)" :
                "Mistral AI n'est pas configuré (clé API manquante)"
    });
});

// Endpoint pour vérifier le statut de Google Vision uniquement
router.get('/google-status', (req, res) => {
    const hasGoogleVisionApiKey = !!process.env.GOOGLE_VISION_API_KEY;
    res.json({
        success: true,
        googleVisionAvailable: hasGoogleVisionApiKey
    });
});

module.exports = router;

File: python_app/routes/bottles.py
"""
Bottles routes for Flask application
Handles all bottle-related API endpoints
"""

from flask import Blueprint, request, jsonify, current_app
import os
import sqlite3
from datetime import datetime
from werkzeug.utils import secure_filename
import base64
import re

bp = Blueprint('bottles', __name__)

# Database path
BASE_DIR = os.path.dirname(os.path.dirname(os.path.dirname(__file__)))
DATABASE_PATH = os.path.join(BASE_DIR, 'python_app', 'data', 'cave.db')
UPLOADS_PATH = os.path.join(BASE_DIR, 'python_app', 'static', 'uploads')

# Ensure uploads directory exists
os.makedirs(UPLOADS_PATH, exist_ok=True)


def get_db_connection():
    """Get a database connection"""
    conn = sqlite3.connect(DATABASE_PATH)
    conn.row_factory = sqlite3.Row
    return conn


def row_to_dict(row):
    """Convert SQLite row to dictionary"""
    if row is None:
        return None
    return {key: row[key] for key in row.keys()}


@bp.route('/', methods=['GET'])
def get_all_bottles():
    """Get all bottles from the database"""
    try:
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute("SELECT * FROM bottles")
        bottles = [row_to_dict(row) for row in cursor.fetchall()]
        conn.close()
        return jsonify(bottles)
    except Exception as e:
        current_app.logger.error(f"Error getting bottles: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@bp.route('/', methods=['POST'])
def save_bottle():
    """Save a bottle to the database"""
    try:
        # Get form data
        if request.content_type and 'multipart/form-data' in request.content_type:
            data = request.form.to_dict()
            # Handle file upload
            if 'photo' in request.files:
                file = request.files['photo']
                if file and file.filename != '':
                    filename = secure_filename(file.filename)
                    unique_filename = f"{datetime.now().timestamp()}-{filename}"
                    file.save(os.path.join(UPLOADS_PATH, unique_filename))
                    data['photo'] = unique_filename
        else:
            data = request.get_json() or {}
        
        # Handle base64 image
        if 'photo' in data and data['photo'] and isinstance(data['photo'], str) and data['photo'].startswith('data:image/'):
            # Extract base64 data
            base64_data = data['photo'].split(',')[1] if ',' in data['photo'] else data['photo']
            try:
                buffer = base64.b64decode(base64_data)
                filename = f"{datetime.now().timestamp()}-camera.jpg"
                file_path = os.path.join(UPLOADS_PATH, filename)
                with open(file_path, 'wb') as f:
                    f.write(buffer)
                data['photo'] = filename
            except Exception as e:
                current_app.logger.error(f"Error decoding base64 image: {e}")
                data['photo'] = ''
        
        # Extract bottle data
        row = int(data.get('row', 0))
        col = int(data.get('col', 0))
        name = data.get('name', '')
        year = int(data.get('year', 0)) if data.get('year') is not None and str(data.get('year')).isdigit() else None
        grapes = data.get('grapes', '')
        region = data.get('region', '')
        drinkFrom = int(data.get('drinkFrom', 0)) if data.get('drinkFrom') is not None and str(data.get('drinkFrom')).isdigit() else None
        drinkTo = int(data.get('drinkTo', 0)) if data.get('drinkTo') is not None and str(data.get('drinkTo')).isdigit() else None
        foodPairing = data.get('foodPairing', '')
        temperature = data.get('temperature', '')
        photo = data.get('photo', '')
        
        conn = get_db_connection()
        cursor = conn.cursor()
        
        # Check if bottle exists at this position
        cursor.execute("SELECT * FROM bottles WHERE row = ? AND col = ?", (row, col))
        existing_bottle = cursor.fetchone()
        
        if existing_bottle:
            # Update existing bottle
            cursor.execute('''
                UPDATE bottles SET
                    name = ?,
                    year = ?,
                    grapes = ?,
                    region = ?,
                    drinkFrom = ?,
                    drinkTo = ?,
                    foodPairing = ?,
                    temperature = ?,
                    photo = ?
                WHERE row = ? AND col = ?
            ''', (name, year, grapes, region, drinkFrom, drinkTo, foodPairing, temperature, photo, row, col))
        else:
            # Insert new bottle
            cursor.execute('''
                INSERT INTO bottles
                (row, col, name, year, grapes, region, drinkFrom, drinkTo, foodPairing, temperature, photo)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            ''', (row, col, name, year, grapes, region, drinkFrom, drinkTo, foodPairing, temperature, photo))
        
        conn.commit()
        conn.close()
        
        return jsonify({'success': True})
    except Exception as e:
        current_app.logger.error(f"Error saving bottle: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@bp.route('/', methods=['DELETE'])
def delete_bottle():
    """Delete a bottle from the database"""
    try:
        row = int(request.args.get('row', 0))
        col = int(request.args.get('col', 0))
        
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute("DELETE FROM bottles WHERE row = ? AND col = ?", (row, col))
        conn.commit()
        conn.close()
        
        return jsonify({'success': True})
    except Exception as e:
        current_app.logger.error(f"Error deleting bottle: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500


@bp.route('/analyze', methods=['POST'])
def analyze_bottle():
    """Analyze a bottle image - fallback implementation"""
    try:
        # This is a fallback implementation
        # In production, you would integrate with Mistral AI and Google Vision
        return jsonify({
            'success': True,
            'bottleInfo': {
                'name': '',
                'year': None,
                'grapes': '',
                'region': '',
                'drinkFrom': None,
                'drinkTo': None,
                'foodPairing': '',
                'temperature': '',
                'analysisMethod': 'Fallback'
            },
            'mistralAvailable': False,
            'googleVisionAvailable': False,
            'extractedText': ''
        })
    except Exception as e:
        current_app.logger.error(f"Error analyzing bottle: {e}")
        return jsonify({
            'success': False,
            'error': str(e),
            'mistralAvailable': False,
            'googleVisionAvailable': False,
            'bottleInfo': {}
        }), 500


@bp.route('/analyze-two-step', methods=['POST'])
def analyze_bottle_two_step():
    """Two-step analysis - fallback implementation"""
    try:
        return jsonify({
            'success': True,
            'bottleInfo': {
                'name': '',
                'year': None,
                'grapes': '',
                'region': '',
                'drinkFrom': None,
                'drinkTo': None,
                'foodPairing': '',
                'temperature': '',
                'analysisMethod': 'Fallback'
            },
            'mistralAvailable': False,
            'googleVisionAvailable': False,
            'requiresManualInput': True,
            'missingFields': {'name': True, 'year': True},
            'partialData': None
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e),
            'mistralAvailable': False,
            'googleVisionAvailable': False,
            'requiresManualInput': True,
            'missingFields': {'name': True, 'year': True},
            'partialData': None
        }), 500


@bp.route('/analyze-text', methods=['POST'])
def analyze_text():
    """Analyze text - fallback implementation"""
    try:
        data = request.get_json() or {}
        name = data.get('name', '')
        year = data.get('year', None)
        grapes = data.get('grapes', '')
        region = data.get('region', '')
        
        return jsonify({
            'success': True,
            'bottleInfo': {
                'name': name,
                'year': year,
                'grapes': grapes,
                'region': region,
                'drinkFrom': None,
                'drinkTo': None,
                'foodPairing': '',
                'temperature': '',
                'analysisMethod': 'Analyse texte manuel'
            }
        })
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@bp.route('/analyze-base64', methods=['POST'])
def analyze_base64():
    """Analyze base64 image - fallback implementation"""
    try:
        data = request.get_json() or {}
        image_data = data.get('image', '')
        
        return jsonify({
            'success': True,
            'bottleInfo': {
                'name': '',
                'year': None,
                'grapes': '',
                'region': '',
                'drinkFrom': None,
                'drinkTo': None,
                'foodPairing': '',
                'temperature': '',
                'analysisMethod': 'Fallback'
            },
            'mistralAvailable': False
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e),
            'mistralAvailable': False,
            'bottleInfo': {}
        }), 500


@bp.route('/recommendations', methods=['GET'])
def get_recommendations():
    """Get wine recommendations - fallback implementation"""
    try:
        occasion = request.args.get('occasion', '')
        food = request.args.get('food', '')
        budget = request.args.get('budget', '')
        
        # Fallback recommendations
        recommendations = []
        
        return jsonify({'success': True, 'recommendations': recommendations})
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@bp.route('/search', methods=['GET'])
def search_wine():
    """Search for a wine by name - fallback implementation"""
    try:
        name = request.args.get('name', '')
        if not name:
            return jsonify({'success': False, 'error': 'Wine name is required'}), 400
        
        # Fallback search
        wine_info = None
        
        return jsonify({'success': wine_info is not None, 'wine': wine_info})
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@bp.route('/reset-db', methods=['POST'])
def reset_database():
    """Reset the database"""
    try:
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute("DELETE FROM bottles")
        cursor.execute("DELETE FROM cave_config")
        cursor.execute("INSERT INTO cave_config (rows, cols, configured) VALUES (5, 10, 0)")
        conn.commit()
        conn.close()
        
        return jsonify({'success': True, 'message': 'Database reset'})
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@bp.route('/chat', methods=['POST'])
def chat():
    """Chat with AI - fallback implementation"""
    try:
        data = request.get_json() or {}
        question = data.get('question', '')
        bottles = data.get('bottles', [])
        
        if not question:
            return jsonify({'success': False, 'error': 'No question provided'}), 400
        
        # Fallback response
        response = f"Je suis votre assistant IA pour la gestion de cave à vin. Vous avez posé la question: {question}"
        
        return jsonify({
            'success': True,
            'response': response,
            'model': 'mistral-tiny'
        })
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@bp.route('/my-bottles', methods=['GET'])
def get_my_bottles():
    """Get bottles from the user's cellar"""
    try:
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute("SELECT * FROM bottles")
        bottles = [row_to_dict(row) for row in cursor.fetchall()]
        conn.close()
        
        return jsonify({'success': True, 'bottles': bottles})
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@bp.route('/wine-pairing', methods=['POST'])
def wine_pairing():
    """Get wine pairing advice - fallback implementation"""
    try:
        data = request.get_json() or {}
        food = data.get('food', '')
        
        if not food:
            return jsonify({'success': False, 'error': 'No food specified'}), 400
        
        # Fallback advice
        advice = f"Pour accompagner {food}, je recommande un vin adapté."
        
        return jsonify({'success': True, 'advice': advice})
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@bp.route('/wine-info', methods=['POST'])
def wine_info():
    """Get wine information - fallback implementation"""
    try:
        data = request.get_json() or {}
        wine_name = data.get('wineName', '')
        
        if not wine_name:
            return jsonify({'success': False, 'error': 'No wine name specified'}), 400
        
        # Fallback info
        info = {
            'name': wine_name,
            'grapes': '',
            'region': '',
            'drinkFrom': None,
            'drinkTo': None,
            'foodPairing': '',
            'temperature': ''
        }
        
        return jsonify({'success': True, 'info': info})
    except Exception as e:
        return jsonify({'success': False, 'error': str(e)}), 500


@bp.route('/mistral-status', methods=['GET'])
def mistral_status():
    """Check Mistral AI status"""
    try:
        import os
        from dotenv import load_dotenv
        load_dotenv()
        
        has_mistral = bool(os.environ.get('MISTRAL_API_KEY'))
        has_google = bool(os.environ.get('GOOGLE_VISION_API_KEY'))
        model = os.environ.get('MISTRAL_MODEL', 'mistral-tiny')
        
        if has_mistral and has_google:
            message = "Mistral AI et Google Vision sont configurés - Analyse complète disponible"
        elif has_mistral:
            message = "Mistral AI configuré, mais Google Vision non configuré (OCR limité)"
        else:
            message = "Mistral AI n'est pas configuré (clé API manquante)"
        
        return jsonify({
            'success': True,
            'mistralAvailable': has_mistral,
            'googleVisionAvailable': has_google,
            'model': model,
            'message': message
        })
    except Exception as e:
        return jsonify({
            'success': True,
            'mistralAvailable': False,
            'googleVisionAvailable': False,
            'model': 'mistral-tiny',
            'message': 'Erreur lors de la vérification du statut'
        })


@bp.route('/google-status', methods=['GET'])
def google_status():
    """Check Google Vision status"""
    try:
        import os
        from dotenv import load_dotenv
        load_dotenv()
        
        has_google = bool(os.environ.get('GOOGLE_VISION_API_KEY'))
        
        return jsonify({
            'success': True,
            'googleVisionAvailable': has_google
        })
    except Exception as e:
        return jsonify({
            'success': True,
            'googleVisionAvailable': False
        })
